File: PathfinderTestPlugin/src/main/java/com/github/ysl3000/bukkit/test/PathfinderTestPlugin.java
package com.github.ysl3000.bukkit.test;

import com.github.ysl3000.bukkit.pathfinding.PathfinderGoalAPI;
import com.github.ysl3000.bukkit.pathfinding.entity.Insentient;
import com.github.ysl3000.bukkit.pathfinding.goals.PathfinderGoalGimmiCookie;
import com.github.ysl3000.bukkit.pathfinding.goals.PathfinderGoalMoveToLocation;
import com.github.ysl3000.bukkit.pathfinding.goals.TalkToStrangers;
import com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal;
import com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector;
import com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderManager;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Creature;
import org.bukkit.entity.Player;
import org.bukkit.entity.Zombie;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.*;
import java.util.concurrent.TimeUnit;

/**
 * Created by ysl3000
 */
public class PathfinderTestPlugin extends JavaPlugin implements Listener {

    private PathfinderManager pathfinderManager;

    private Map<String, ICommand> commandMap = new HashMap<>();

    @Override
    public void onEnable() {


        this.pathfinderManager = PathfinderGoalAPI.getAPI();

        commandMap.put("chat", new Chat(pathfinderManager));
        commandMap.put("cookie", new DeliverCookie(pathfinderManager));
        commandMap.put("move", new MoveToLocation(pathfinderManager));
        commandMap.put("print", new PrintGoal(pathfinderManager));

        Bukkit.getServer().getPluginManager().registerEvents(this, this);
    }


    @EventHandler
    public void onChat(AsyncPlayerChatEvent event) {

        Player player = event.getPlayer();

        List<String> args = Arrays.asList(event.getMessage().split(" "));

        if (args.size() > 0) {

            String command = args.get(0);

            ICommand iCommand = commandMap.get(command);

            List<String> param = new ArrayList<>();

            if (args.size() > 1) {
                param = args.subList(1, args.size());
            }


            List<String> finalParam = param;
            Bukkit.getScheduler().runTask(this, () -> {

                if (iCommand != null)
                    iCommand.execute(player, finalParam);

            });
        }
    }

    private class Chat implements ICommand {


        private PathfinderManager pathfinderManager;

        private Chat(PathfinderManager pathfinderManager) {
            this.pathfinderManager = pathfinderManager;
        }


        @Override
        public boolean execute(Player player, List<String> args) {

            Creature creature = player.getWorld().spawn(player.getLocation(), Zombie.class);
            Insentient pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature);

            if (pathfinderGoalEntity != null) {
                PathfinderGoalSelector pathfinderGoalSelector = pathfinderGoalEntity.getGoalSelector();
                pathfinderGoalSelector.clearGoals();
                pathfinderGoalEntity.getTargetSelector().clearGoals();

                pathfinderGoalSelector.addPathfinderGoal(0, new TalkToStrangers(pathfinderGoalEntity, args, TimeUnit.SECONDS.toMillis(20)));
            }
            return true;
        }
    }

    private class DeliverCookie implements ICommand {

        private PathfinderManager pathfinderManager;

        private DeliverCookie(PathfinderManager pathfinderManager) {
            this.pathfinderManager = pathfinderManager;
        }

        @Override
        public boolean execute(Player player, List<String> args) {

            Creature creature = player.getWorld().spawn(new Location(player.getWorld(), 235, 70, 246), Zombie.class);
            Insentient pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature);

            if (pathfinderGoalEntity != null) {
                PathfinderGoalSelector pathfinderGoalSelector = pathfinderGoalEntity.getGoalSelector();
                pathfinderGoalSelector.clearGoals();
                pathfinderGoalEntity.getTargetSelector().clearGoals();

                pathfinderGoalSelector.addPathfinderGoal(0, new PathfinderGoalGimmiCookie(pathfinderGoalEntity, creature));
            }

            return true;
        }
    }

    private class MoveToLocation implements ICommand {

        private PathfinderManager pathfinderManager;

        private MoveToLocation(PathfinderManager pathfinderManager) {
            this.pathfinderManager = pathfinderManager;
        }

        @Override
        public boolean execute(Player player, List<String> args) {

            Creature creature = player.getWorld().spawn(player.getLocation(), Zombie.class);
            Insentient pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature);

            if (pathfinderGoalEntity != null) {
                PathfinderGoalSelector pathfinderGoalSelector = pathfinderGoalEntity.getGoalSelector();
                pathfinderGoalSelector.clearGoals();
                pathfinderGoalEntity.getTargetSelector().clearGoals();

                pathfinderGoalSelector.addPathfinderGoal(0,
                        new PathfinderGoalMoveToLocation(
                                pathfinderGoalEntity, new Location(player.getWorld(), 235, 70, 246),
                                2, 0)
                );
            }
            return true;
        }
    }

    private class PrintGoal implements ICommand {
        private final PathfinderManager pathfinderManager;

        public PrintGoal(PathfinderManager pathfinderManager) {

            this.pathfinderManager = pathfinderManager;
        }

        @Override
        public boolean execute(Player p, List<String> args) {
            Creature creature = p.getWorld().spawn(p.getLocation(), Zombie.class,zombie -> {
                Insentient pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(zombie);

                if (pathfinderGoalEntity != null) {
                    PathfinderGoalSelector pathfinderGoalSelector = pathfinderGoalEntity.getGoalSelector();
                    pathfinderGoalEntity.getTargetSelector().clearGoals();
                    pathfinderGoalSelector.clearGoals();

                    pathfinderGoalSelector.addPathfinderGoal(0,
                            new PathfinderGoalPrint()
                    );
                }


            });

            return true;
        }

        private class PathfinderGoalPrint implements PathfinderGoal {

            private boolean shE = true;
            private boolean shT = true;

            /**
             * Whether the pathfinder goal should commence execution or not
             *
             * @return true if should execute
             */
            @Override
            public boolean shouldExecute() {

                System.out.println("called should execute");

                return shE = !shE;
            }

            /**
             * Whether the goal should Terminate
             *
             * @return true if should terminate
             */
            @Override
            public boolean shouldTerminate() {
                System.out.println("Called should terminate");
                return shT = !shT;
            }

            /**
             * Runs initially and should be used to setUp goalEnvironment
             * Condition needs to be defined thus your code in it isn't called
             */
            @Override
            public void init() {

                System.out.println("Called Init");

            }

            /**
             * Is called when {@link #shouldExecute()} returns true
             */
            @Override
            public void execute() {

                System.out.println("Called execute");

            }

            /**
             * Reset the pathfinder AI pack to its initial state
             * <p>
             * Is called when {@link #shouldExecute()} returns false
             */
            @Override
            public void reset() {

                System.out.println("Called reset");

            }
        }
    }
}

File: PathfinderTestPlugin/src/main/java/com/github/ysl3000/bukkit/test/PathfinderTestPlugin.kt
package com.github.ysl3000.bukkit.test

import com.github.ysl3000.bukkit.pathfinding.PathfinderGoalAPI
import com.github.ysl3000.bukkit.pathfinding.entity.Insentient
import com.github.ysl3000.bukkit.pathfinding.goals.PathfinderGoalGimmiCookie
import com.github.ysl3000.bukkit.pathfinding.goals.PathfinderGoalMoveToLocation
import com.github.ysl3000.bukkit.pathfinding.goals.TalkToStrangers
import com.github.ysl3000.bukkit.pathfinding.goals.notworking.PathfinderGoalFollowEntity
import com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal
import com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderManager
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.entity.Creature
import org.bukkit.entity.Player
import org.bukkit.entity.Zombie
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.plugin.java.JavaPlugin

import java.util.*
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.TimeUnit

/**
 * Created by ysl3000
 */
class PathfinderTestPlugin : JavaPlugin(), Listener {

    private val commandMap = HashMap<String, ICommand>()

    override fun onEnable() {

        val pathfinderManager = PathfinderGoalAPI.api

        commandMap["chat"] = Chat(pathfinderManager)
        commandMap["cookie"] = DeliverCookie(pathfinderManager)
        commandMap["move"] = MoveToLocation(pathfinderManager)
        commandMap["follow"] = FollowOwner(pathfinderManager)
        commandMap["print"] = PrintGoal(pathfinderManager)

        Bukkit.getServer().pluginManager.registerEvents(this, this)
    }


    @EventHandler
    fun onChat(event: AsyncPlayerChatEvent) {

        val player = event.player

        val args = Arrays.asList(*event.message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())

        if (!args.isEmpty()) {

            val command = args[0]

            val iCommand = commandMap[command]

            var param: List<String> = ArrayList()

            if (args.size > 1) {
                param = args.subList(1, args.size)
            }

            val finalParam = param
            Bukkit.getScheduler().runTask(this) {

                iCommand?.execute(player, finalParam)

            }
        }
    }

    private inner class Chat private constructor(private val pathfinderManager: PathfinderManager) : ICommand {


        override fun execute(player: Player, args: List<String>): Boolean {

            val creature = player.world.spawn(player.location, Zombie::class.java)
            val pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature)

            pathfinderGoalEntity.clearPathfinderGoals()
            pathfinderGoalEntity.clearTargetPathfinderGoals()
            pathfinderGoalEntity.addPathfinderGoal(0,
                    TalkToStrangers(pathfinderGoalEntity, args, TimeUnit.SECONDS.toMillis(20)))
            return true
        }
    }

    private inner class DeliverCookie private constructor(private val pathfinderManager: PathfinderManager) : ICommand {

        override fun execute(player: Player, args: List<String>): Boolean {

            val rand = ThreadLocalRandom.current()

            val spawn = player.location.add(rand.nextDouble(10.0), 0.0, rand.nextDouble(10.0))


            val creature = player.world
                    .spawn(spawn, Zombie::class.java)
            val pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature)

            pathfinderGoalEntity.clearPathfinderGoals()
            pathfinderGoalEntity.clearTargetPathfinderGoals()
            pathfinderGoalEntity
                    .addPathfinderGoal(0, PathfinderGoalGimmiCookie(pathfinderGoalEntity))

            return true
        }
    }

    private inner class MoveToLocation private constructor(private val pathfinderManager: PathfinderManager) : ICommand {

        override fun execute(player: Player, args: List<String>): Boolean {

            val creature = player.world.spawn(player.location, Zombie::class.java)
            val pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature)
            pathfinderGoalEntity.clearPathfinderGoals()
            pathfinderGoalEntity.clearTargetPathfinderGoals()
            pathfinderGoalEntity.addPathfinderGoal(0,
                    PathfinderGoalMoveToLocation(
                            pathfinderGoalEntity, Location(player.world, 235.0, 70.0, 246.0),
                            2.0, 0.0)
            )

            return true
        }
    }

    private inner class FollowOwner private constructor(private val pathfinderManager: PathfinderManager) : ICommand {

        override fun execute(player: Player, args: List<String>): Boolean {

            val creature = player.world.spawn(player.location, Zombie::class.java)
            val pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature)
            pathfinderGoalEntity.clearPathfinderGoals()
            pathfinderGoalEntity.clearTargetPathfinderGoals()
            pathfinderGoalEntity.addPathfinderGoal(0,
                    PathfinderGoalFollowEntity(pathfinderGoalEntity, player, 5.0, 0.2)
            )

            return true
        }
    }


    private inner class PrintGoal(private val pathfinderManager: PathfinderManager) : ICommand {

        override fun execute(p: Player, args: List<String>): Boolean {
            val creature = p.world.spawn(p.location, Zombie::class.java)

            val pathfinderGoalEntity = this.pathfinderManager.getPathfindeGoalEntity(creature)
            pathfinderGoalEntity.clearPathfinderGoals()
            pathfinderGoalEntity.clearTargetPathfinderGoals()
            pathfinderGoalEntity.addPathfinderGoal(0,
                    PathfinderGoalPrint()
            )

            return true
        }

        private inner class PathfinderGoalPrint : PathfinderGoal {

            private var shE = true
            private var shT = true

            override fun shouldExecute(): Boolean {
                println("called should execute")
                return shE = !shE
            }

            override fun shouldTerminate(): Boolean {
                println("Called should terminate")
                return shT = !shT
            }

            override fun init() {
                println("Called Init")
            }

            override fun execute() {
                println("Called execute")
            }

            override fun reset() {
                println("Called reset")
            }
        }
    }
}
